src/librustc_mir/transform/validate.rs

```diff
@@ -23,12 +23,14 @@ enum EdgeKind {
 pub struct Validator {
     /// Describes at which point in the pipeline this validation is happening.
     pub when: String,
+    pub mir_phase: MirPhase,
```

Comment:
This should make clear how the phase is used. I guess it means that we are validating for the dialect that should be in place *before* this phase?

Reply:
That means if `Optimized` changed the dialect, we'd never check that. Even at the end of `DropElab` we would not check that aggregates are gone, since we run validation with the `DropElab` phase which still allows aggregrates.

Reply:
I added some documentation, but yes, we don't check if you're changing the dialect. Each phase expects a certain dialect, and we're ensuring that this holds up. This means we may get errors about dialect misuse one optimization too late, but it is pretty obvious from the error what's wrong.

Reply:
I can live with this if it is properly explained: validation checks MIR for the dialect expected *as input* by the current phase.

> This means we may get errors about dialect misuse one optimization too late, but it is pretty obvious from the error what's wrong.

It also means we never check the dialect specific to the last phase. Good thing that right now, the last phase doesn't change dialects.

Reply:
There's an off-by-one between what the comments in the `MirPhase` enum say (dialect *after* this phase), and what validation checks (dialect *before* this phase). That's the confusing bit that at least needs through documentation.

Reply:
I wrote some documentation in the appropriate places, wdyt?

Reply:
LGTM. I can't say if this would have been enough without the discussion we've had, but also I can't think of obvious local improvements, so let's go with this for now.